crates/storage/src/connection/pruning.rs

```diff
@@ -0,0 +1,347 @@
+        latest: BlockNumber,
+        num_blocks_kept: u64,
+    ) -> anyhow::Result<()> {
+        let block_to_prune = latest
+            .checked_sub(num_blocks_kept + 1)
+            .expect("Blockchain pruning should not start before there are blocks to prune");
+
+        tracing::info!(%block_to_prune, "Running blockchain pruning");
+        let start = std::time::Instant::now();
+        self.prune_transaction_data(block_to_prune)
+            .context("Pruning transaction data")?;
+        self.prune_block_data(block_to_prune)
+            .context("Pruning block data")?;
+        tracing::debug!(pruned_block=%block_to_prune, elapsed=?start.elapsed(), "Blockchain pruning done");
+
+        Ok(())
+    }
+
+    fn prune_transaction_data(&self, block: BlockNumber) -> anyhow::Result<()> {
+        let mut transaction_stmt = self.inner().prepare_cached(
+            r"
+            DELETE FROM transactions
+            WHERE block_number = ?
+            ",
+        )?;
+        let mut transaction_hashes_stmt = self.inner().prepare_cached(
+            r"
+            DELETE FROM transaction_hashes
+            WHERE block_number = ?
+            ",
+        )?;
+        transaction_stmt.execute(params![&block])?;
+        transaction_hashes_stmt.execute(params![&block])?;
+
+        Ok(())
+    }
+
+    fn prune_block_data(&self, block: BlockNumber) -> anyhow::Result<()> {
+        let mut block_is_prunable_stmt = self.inner().prepare_cached(
+            r"
+            SELECT EXISTS (
+                SELECT 1
+                FROM block_headers
+                WHERE number = ?
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM contract_updates
+                    WHERE block_number = block_headers.number
+                )
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM nonce_updates
+                    WHERE block_number = block_headers.number
+                )
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM storage_updates
+                    WHERE block_number = block_headers.number
+                )
+            ) AS has_no_updates
+            ",
+        )?;
+
+        let prunable = block_is_prunable_stmt
+            .query_row(params![&block], |row| row.get::<_, bool>(0))
+            .context(format!("Checkig if block {block} is prunable"))?;
+
+        // A block without any state update cannot make any other state updates
+        // obsolete. Moreover, it can immediately be pruned since there are no updates
+        // to make obsolete on it.
+        if prunable {
+            self.prune_block(&block)?;
+            return Ok(());
+        }
+
+        let mut contract_updates_select_stmt = self.inner().prepare_cached(
+            r"
+            SELECT contract_address
+            FROM contract_updates
+            WHERE block_number = :last_kept_block
+            ",
+        )?;
```

Comment:
Pruning `contract_updates` might be a bit more complicated: state diff reconstruction depends on the last _two_ updates (see [this](https://github.com/eqlabs/pathfinder/blob/5f9b6bb977f606d0483543128fd19ffc5e498343/crates/storage/src/connection/state_update.rs#L427) part of `.state_update()`).